controllers/awscluster_controller.go

```diff
@@ -68,9 +69,45 @@ var (
 // AWSClusterReconciler reconciles a AwsCluster object.
 type AWSClusterReconciler struct {
 	client.Client
-	Recorder         record.EventRecorder
-	Endpoints        []scope.ServiceEndpoint
-	WatchFilterValue string
+	Recorder              record.EventRecorder
+	ec2ServiceFactory     func(scope.EC2Scope) services.EC2MachineInterface
+	networkServiceFactory func(scope.ClusterScope) services.NetworkInterface
+	elbServiceFactory     func(scope.ELBScope) services.ELBInterface
+	securityGroupFactory  func(scope.ClusterScope) services.SecurityGroupInterface
+	Endpoints             []scope.ServiceEndpoint
+	WatchFilterValue      string
+}
+
+// getEC2Service factory func is added for testing purpose so that we can inject mocked EC2Service to the AWSClusterReconciler.
+func (r *AWSClusterReconciler) getEC2Service(scope scope.EC2Scope) services.EC2MachineInterface {
```

Comment:
I'm wondering if we could change the pattern we are using here slightly so that we don't have these extra functions and instead we handle via arguments to `New` . So, something like this (I haven't tested this and will need changing/investigating):

```go
type NewClusterReconcilerInput struct {
	Manager ctrl.Manager
	WatchFilterValue string
	Endpoints []scope.ServiceEndpoint
}

type NewClusterReconcilerServices struct {
	EC2ServiceFactory     func(scope.EC2Scope) services.EC2MachineInterface
	NetworkServiceFactory func(scope.ClusterScope) services.NetworkInterface
	ElbServiceFactory     func(scope.ELBScope) services.ELBInterface
	SecurityGroupFactory  func(scope.ClusterScope) services.SecurityGroupInterface
}

func NewClusterReconciler(input NewClusterReconcilerInput) reconcile.Reconciler {
	return &awsClusterReconciler{
		Client:           input.Manager.GetClient(),
		Recorder:         input.Manager.GetEventRecorderFor("awscluster-controller"),
		Endpoints:        input.Endpoints,
		WatchFilterValue: input.WatchFilterValue,

		ec2ServiceFactory: ec2.NewService,
		elbServiceFactory: elb.NewService,
		networkServiceFactory: network.NewService,
		securityGroupFactory: securitygroup.NewService,
	}
}

func NewClusterReconcilerWithServices(input NewClusterReconcilerInput, svcs NewClusterReconcilerServices ) reconcile.Reconciler {
	return &awsClusterReconciler{
		Client:           input.Manager.GetClient(),
		Recorder:         input.Manager.GetEventRecorderFor("awscluster-controller"),
		Endpoints:        input.Endpoints,
		WatchFilterValue: input.WatchFilterValue,

		ec2ServiceFactory: svcs.EC2ServiceFactory,
		elbServiceFactory: svcs.ElbServiceFactory,
		networkServiceFactory: svcs.NetworkServiceFactory,
		securityGroupFactory: svcs.SecurityGroupFactory,
	}
}

// awsClusterReconciler reconciles a AwsCluster object.
type awsClusterReconciler struct {
	client.Client
	Recorder         record.EventRecorder
	Endpoints        []scope.ServiceEndpoint
	WatchFilterValue string

	ec2ServiceFactory     func(scope.EC2Scope) services.EC2MachineInterface
	networkServiceFactory func(scope.ClusterScope) services.NetworkInterface
	elbServiceFactory     func(scope.ELBScope) services.ELBInterface
	securityGroupFactory  func(scope.ClusterScope) services.SecurityGroupInterface
}
```

And maybe adopt functional options, which could be something along these lines (again not tested and would need to be changed/thought about):

```go
type ReconcilerOption func(*AWSReconciler)

type EC2ServiceFactory func(scope.EC2Scope) services.EC2MachineInterface

func WithEC2ServiceFactory(fn EC2ServiceFactory) ReconcilerOption {
	return func(r *AWSReconciler) {
		r.EC2ServiceFactory  = fn
	}
}


func NewClusterReconciler(input NewClusterReconcilerInput, opt ...ReconcilerOption) reconcile.Reconciler {
	clusterReconciler := &awsClusterReconciler{
		Client:           input.Manager.GetClient(),
		Recorder:         input.Manager.GetEventRecorderFor("awscluster-controller"),
		Endpoints:        input.Endpoints,
		WatchFilterValue: input.WatchFilterValue,

		EC2ServiceFactory: ec2.NewService,
		ELBServiceFactory: elb.NewService,
		NetworkServiceFactory: network.NewService,
		SecurityGroupFactory: securitygroup.NewService,
	}

	for _, opt := range opt {
		opt(clusterReconciler)
	}

	return clusterReconciler
}
```

Reply:
@richardcase 
A question to understand better:
Reconcilers are created in main and we do not have the session created there yet. How do we create `ec2.NewService` in this case when we call NewClusterReconciler()?

AWSMachineReconciler have serviceFactories in its struct, which is better in the sense that we don't have to call `ec2.NewService()` multiple times for reconcileNormal() and reconcileDelete().

Reply:
Do you want to track this in a new issue? We will probably need to do changes everywhere if we want to proceed. So out of scope of this PR.

Reply:
> Reconcilers are created in main and we do not have the session created there yet. How do we create ec2.NewService in this case when we call NewClusterReconciler()?

We don't create an instance of the services in or when calling `NewClusterReconciler` but instead we pass in functions. So in this example, we are passing in the function `ec2.NewService` and not calling it `ec2.NewService(scope)`. The function we pass in should adhere to the same factory function definition. We can then call the function later when we have the scope.  But it also allows us to override the factory function when we create an new instance of the reconciler in the tests. And if we don't pass in an override it can default to what we want outside of testing, `ec2.NewService` in this example.

Instead of having logic like this in the controllers:

```go
func (r *AWSClusterReconciler) getEC2Service(scope scope.EC2Scope) services.EC2MachineInterface {
	if r.ec2ServiceFactory != nil {
		return r.ec2ServiceFactory(scope)
	}
	return ec2.NewService(scope)
}
```

it feels to me that we should handle this during creation of the controller (i.e. via a concreational pattern) and so anywhere we need an ec2 service in the reconciler we just go `r.ec2ServiceFactory(scope)`.

Reply:
> Do you want to track this in a new issue? We will probably need to do changes everywhere if we want to proceed with the change. So out of scope of this PR.

These changes are being made to aid testability so i would argue its in scope but i should've picked up on this earlier and as you point out it requires changes across a number of controllers. 

So, i will create a new issue and follow up after these testing PRs merge :+1:

Reply:
@richardcase @sedefsavas so shall we move this discussion to a new issue and go ahead with the PR changes?

Reply:
Yes i will create the new issue with the above :)

Reply:
cc @pydctw